pkg/platform/api/graphql/model/buildplanner/buildExpression.go

```diff
@@ -56,26 +56,63 @@ type Requirement struct {
 
 type VersionRequirement map[Comparator]string
 
-type BuildExpression map[string]interface{}
-
-func NewBuildExpression(data []byte) (BuildExpression, error) {
-	var bx BuildExpression
+type BuildExpression struct {
+	expression       map[string]interface{}
+	solveNode        *map[string]interface{}
+	requirementsNode []interface{}
+}
 
-	err := json.Unmarshal(data, &bx)
+func NewBuildExpression(data []byte) (*BuildExpression, error) {
+	expression := make(map[string]interface{})
+	err := json.Unmarshal(data, &expression)
 	if err != nil {
 		return nil, errs.Wrap(err, "Could not unmarshal JSON")
 	}
 
-	return bx, nil
-}
+	solveNode, err := getSolveNode(expression)
+	if err != nil {
+		return nil, errs.Wrap(err, "Could not get solve node")
+	}
 
-func (bx BuildExpression) Requirements() ([]Requirement, error) {
-	requirementsNode, err := getRequirementsNode(bx)
+	requirementsNode, err := getRequirementsNode(expression)
 	if err != nil {
 		return nil, errs.Wrap(err, "Could not get requirements node")
 	}
 
-	requirementsData, err := json.Marshal(requirementsNode)
+	err = validateRequirements(requirementsNode)
+	if err != nil {
+		return nil, errs.Wrap(err, "Requirements in BuildExpression are invalid")
+	}
+
+	return &BuildExpression{
+		expression:       expression,
+		solveNode:        &solveNode,
+		requirementsNode: requirementsNode,
+	}, nil
+}
+
+func validateRequirements(requirements []interface{}) error {
+	for _, requirement := range requirements {
+		r, ok := requirement.(map[string]interface{})
+		if !ok {
+			return errs.New("Requirement in BuildExpression is malformed")
+		}
+		_, ok = r[RequirementNameKey]
+		if !ok {
+			return errs.New("Requirement in BuildExpression is missing name field")
```

Comment:
Nit: it may be helpful for debugging (and rollbar) to put a `%v` for r in the error message so that we know what requirement we're looking at. If we have to hunt through a wall of JSON to find the bad requirement, that would not be fun.